### DISCUSSION TITLE
Installation Issue - MET-10.0.0 - zlib.h

### DISCUSSION THREAD

**Discussion by spence98** [dtcenter/METplus#1179] [Installation] [answered]
Hello! I am trying to install MET-10.0.0. I have configured it, but when I try to run 'make' I am getting the following error:

In file included from vx_render.h(31),
                 from render_pbm.cc(23):
flate_filter.h(20): catastrophic error: cannot open source file "zlib.h"
  #include <zlib.h>
                   ^

compilation aborted for render_pbm.cc (code 4)
make[4]: *** [libvx_render_a-render_pbm.o] Error 4
make[4]: Leaving directory `/raid-08/LS/spence98/met-10.0.0/src/libcode/vx_render'
make[3]: *** [all-recursive] Error 1
make[3]: Leaving directory `/raid-08/LS/spence98/met-10.0.0/src/libcode'
make[2]: *** [all-recursive] Error 1
make[2]: Leaving directory `/raid-08/LS/spence98/met-10.0.0/src'
make[1]: *** [all-recursive] Error 1
make[1]: Leaving directory `/raid-08/LS/spence98/met-10.0.0'
make: *** [all] Error 2

I am unsure of why this error is occurring, since I do have the zlib.h file present. Any help would be greatly appreciated! I've attached the config.log file in case that helps.

[config.log](https://github.com/dtcenter/METplus/files/7239409/config.log)

I will say, I did successfully install the 9.1.1 version about nine months ago and it never had this particular error in the installation process. 

Thanks,
Michelle

**Comment by jprestop** [comment, marked as answer]
Hi Michelle.

Could you please try adding the path to the CPPFLAGS variable with "-I"?

So instead of the following:

```
CPPFLAGS=' -I${MET_NETCDF}/include -I${MET_HDF5}/include -I/raid-08/LS/spence98/software/grib2/g2clib_install/g2clib-1.6.0 -I${MET_GSL}/include -DBLOCK4 -DWITH\
_GRIB2 -DMET_BASE="\"$(pkgdatadir)\""'
```
You would have the following, replacing \<path to zlib.h\> with the actual path to zlib.h:

```
CPPFLAGS='-I<path to zlib> -I${MET_NETCDF}/include -I${MET_HDF5}/include -I/raid-08/LS/spence98/software/grib2/g2clib_install/g2clib-1.6.0 -I${MET_GSL}/include -DBLOCK4 -DWITH\
_GRIB2 -DMET_BASE="\"$(pkgdatadir)\""'
```

Please give that a try and let us know how it goes.

**Reply by spence98** [reply]
Good afternoon,

Thank you for your quick response! I did as you suggested and that helped me to get past that error. However, when I run make now I am getting a different error:

ld: cannot find -ljasper
ld: cannot find -lpng
ld: cannot find -lz
make[5]: *** [ensemble_stat] Error 1
make[5]: Leaving directory `/raid-08/LS/spence98/met-10.0.0/src/tools/core/ensemble_stat'
make[4]: *** [all-recursive] Error 1
make[4]: Leaving directory `/raid-08/LS/spence98/met-10.0.0/src/tools/core'
make[3]: *** [all-recursive] Error 1
make[3]: Leaving directory `/raid-08/LS/spence98/met-10.0.0/src/tools'
make[2]: *** [all-recursive] Error 1
make[2]: Leaving directory `/raid-08/LS/spence98/met-10.0.0/src'
make[1]: *** [all-recursive] Error 1
make[1]: Leaving directory `/raid-08/LS/spence98/met-10.0.0'
make: *** [all] Error 2

I've attached an updated config.log file.

[config.log](https://github.com/dtcenter/METplus/files/7246139/config.log)

My understanding of this error is that MET is unable to find where my jasper and libpng libraries are. I'm not sure if I am correct in this line of thinking though. I did try adding another -I<path to libpng> to the CPPFLAGS to see if that would take care of it (at least for the 'ld: cannot find -lpng' portion, but that didn't seem to make a difference.

Michelle

**Reply by jprestop** [reply]
Hi Michelle. 

Thank you for letting us know that you were able to get past that error and for sharing the new errors.  

We can use CPPFLAGS to specify the include directories.  In this case, we need to use LDFLAGS to specify the library files.  Please find the locations of libjasper.a (or libjasper.so), libpng.a, and libz.a and will need to add them to the LDFLAGS variable with "-L"?

So instead of the following:

```
LDFLAGS=' -L${MET_NETCDF}/lib -L${MET_HDF5}/lib -L/raid-08/LS/spence98/software/grib2/g2clib_install/g2clib-1.6.0 -L${MET_GSL}/lib -L/raid-11/LS/evans36/software/bufrlib'
```
You would have the following, replacing all of the \<path to ... \> with the actual paths to the library files:

```
LDFLAGS=' -L<path to libjasper.a> -L<path to libpng.a> -L<path to libz.a> -L${MET_NETCDF}/lib -L${MET_HDF5}/lib -L/raid-08/LS/spence98/software/grib2/g2clib_install/g2clib-1.6.0 -L${MET_GSL}/lib -L/raid-11/LS/evans36/software/bufrlib'
```

Please give that a try and let us know how it goes.

**Reply by spence98** [reply]
Hello again,

Thank you for explaining the difference between those two steps! I did make those changes and that got me one step closer to installing (I think). I do have another error though:

//raid-11/LS/evans36/software/grib2-new/lib/libpng16.so.16: undefined reference to `inflateValidate@ZLIB_1.2.9'
make[5]: *** [tc_dland] Error 1
make[5]: Leaving directory `/raid-08/LS/spence98/met-10.0.0/src/tools/tc_utils/tc_dland'

I am not sure exactly why it is trying to reference my advisors grib2-new/lib folder for the libpng16.so.16 file since I have specified finding the libpng.a file within my own directories with the LDFLAGS command. 

I'm wondering if I need to configure/install the rest of the necessary libraries myself (instead of using my advisor's netcdf, gls, bufrlib, and hdf5 libraries). I currently have grib2, jasper, libpng, and zlib configured and installed within my own working directory, but I point to my advisor's for the others. 

Updated config log:
[config.log](https://github.com/dtcenter/METplus/files/7247632/config.log)

Thank you for your continued help!
Michelle

**Reply by jprestop** [reply]
HI Michelle.  Thank you for letting me know that some progress was made.  Could you please also attach your make.log file so I can take a look at it?

**Reply by spence98** [reply]
Hi! Yes, here it is, sorry about that!

[make_install.log](https://github.com/dtcenter/METplus/files/7247654/make_install.log)

**Reply by jprestop** [reply]
I am also perplexed why it is trying to access the grib2-new/lib folder for the png library.  I don't see it listed in the PATH (or anywhere else) in your config.log file, nor is it in your make.log file except where the error appears.   

> I'm wondering if I need to configure/install the rest of the necessary libraries myself (instead of using my advisor's netcdf, gls, bufrlib, and hdf5 libraries). I currently have grib2, jasper, libpng, and zlib configured and installed within my own working directory, but I point to my advisor's for the others.

That could be necessary.  However, let's have you try one thing first.  Please set the following in your environment, then please run "make clean" and reconfigure and remake MET:

```
export LIBS='-lvx_data2d_grib2 -lgrib2c -ljasper -lpng -lz'
```

Please let us know how it goes.  If the build is not successful and you end up needing to rebuild the other libraries, I have a script that you could use to build MET and all of its dependencies, if that is something you are interested in using.  You can find the script and directions for it on the [MET downloads page](https://dtcenter.org/community-code/model-evaluation-tools-met/download) under "Sample Script For Compiling External Libraries And MET".

**Reply by spence98** [reply]
Hello!

So, an update. I spoke with my advisor to see if he had any ideas as to why it would be referencing his libpng files and it turns out gsl was pointing towards it in the background. 

So, I installed gsl within my own directory and fixed the necessary paths within my .bashrc and that solved the issue! It has now successfully been installed. Thank you for all of your help in this process!

Michelle

**Reply by jprestop** [reply]
Hi Michelle.  That's great news!  Thank you so much for following up.  

At this point I'm going to LOCK this conversation to prevent future posts. Our team has decided that once discussions have been answered, and the answer has been confirmed, we'll lock them. We want to encourage users to ask new questions in new discussions rather than posting to old ones. Hopefully that'll make the questions/answers easier for other users to follow.

So if/when more issues/questions arise, please feel free to start a new discussion.
